### ISSUE TITLE
Is it possible to break down a force into individual interactions?

### ISSUE THREAD

**Issue by Gareth-elliott** [openmm/openmm#3409]
Hi OpenMM Developers,

I have been converting some parameters from tinker to openmm and (especially after the fix in #3376 ) this has been mostly successful. I did have an issue today with the conversion of one particular molecule (ethanol). All the terms are fine, except the stretch-bend term:

TINKER:
 Energy Component Breakdown :           Kcal/mole        Interactions
 Bond Stretching                           0.9611                8
 Angle Bending                             0.2219               13
 **Stretch-Bend                              0.0186                9**
 Torsional Angle                           0.0011               12
 Van der Waals                             0.8595               15
 Atomic Multipoles                        -0.9873               15
 Polarization                             -0.2529               15

OPENMM:
AmoebaBond = 0.9610670410837722
AmoebaAngle = 0.22193213789688226
AmoebaInPlaneAngle = 0.0
**AmoebaStretchBend = -0.00495009064193708**
PeriodicTorsionForce = 0.0010703517696935574
AmoebaVdwForce = 0.8595058606240733
AmoebaMultipoleForce = -1.2402446877061468
CMMotionRemover = 0.0

My question is if it is possible to decompose the stretch-bend force into the individual interactions to make sure that there are the same number and that everything is assigned correctly (or if there is a better way to check). Something like the output from tinker ANALYZE:

 Individual Stretch-Bend Interactions :

 Type                  Atom Names                  dSB 1     dSB 2      Energy

 StrBend         4-H        1-C        5-C        0.0067    0.0043      0.0023
 StrBend         1-C        5-C        6-H       -0.0097   -0.0114     -0.0056
 StrBend         7-H        5-C        8-O       -0.0250    0.0085      0.0076
 StrBend         5-C        8-O        9-H        0.0121   -0.0233      0.0099
 StrBend         2-H        1-C        5-C        0.0059    0.0037      0.0020
 StrBend         3-H        1-C        5-C        0.0059    0.0037      0.0020
 StrBend         1-C        5-C        8-O       -0.0039    0.0015     -0.0016
 StrBend         1-C        5-C        7-H       -0.0097   -0.0114     -0.0056
 StrBend         6-H        5-C        8-O       -0.0250    0.0085      0.0076

Thanks,
Gareth

**Comment by peastman** [comment]
You can print out all stretch-bend terms and their parameters by doing something like this:

```python
for force in system.getForces():
  if force.getName() == 'AmoebaStretchBend':
    for i in range(force.getNumBonds()):
      print(force.getBondParameters(i))
```

If you want to get the energy of each one individually, that's harder.  There's no public API for querying it, but if you don't mind adding debugging code and compiling from source, I can tell you exactly what to do to make it print them.

**Comment by Gareth-elliott** [comment]
Thanks for the quick reply @peastman!
I think I might need to get the energies, it appears as though the correct number of terms is being allocated, and the angles and distances match up.

**Comment by Gareth-elliott** [comment]
Thanks @peastman :-)

I still don't know why I'm getting a different energy from this (when multiple other molecules I have converted have worked). The energy values I get are shown in the table below, with the bottom row indicating the sums of each (with agree with the initial post).

I tried to understand it a bit more by looking at how it gets added by the forcefield class, and from what I understand it needs both the bonding and angle terms to be defined first as it relies on them. 

Both of those appear to be fine from the energy comparison (initial post), and comparing the parameter assignments from Tinker and OpenMM they both are assigning the same initial distances and angles to the stretch-bend force. The force constants also appear to match (the analyze program in Tinker only reports one value, it appears to be the arithmetic mean of the two parameters, and hence OpenMM appears to assign the same as Tinker).

I'm not sure what else I can do to try and debug this. I'm happy to keep looking into it. I'll attach some files in case you want to have a look. 

[ethanol_analyze_full.out.txt](https://github.com/openmm/openmm/files/7860068/ethanol_analyze_full.out.txt) <- output from Tinker's ANALYZE "analyze ethanol GPEALDMVC > ethanol_analyze_full.out"
[ethanol_from_amoeba09.prm.txt](https://github.com/openmm/openmm/files/7860069/ethanol_from_amoeba09.prm.txt) <- the selection of parameters from amoeba09.prm that are required for ethanol
[ethanol.pdb.txt](https://github.com/openmm/openmm/files/7860070/ethanol.pdb.txt) <- pdb
[ethanol.xml.txt](https://github.com/openmm/openmm/files/7860071/ethanol.xml.txt) <- converted .xml forcefield file
[ethanol.xyz.txt](https://github.com/openmm/openmm/files/7860096/ethanol.xyz.txt) <- .xyz file for Tinker
[computeEnergies.py.txt](https://github.com/openmm/openmm/files/7860105/computeEnergies.py.txt) <- script for getting the energies in OpenMM



Tinker | OpenMM | OpenMM - Tinker
-- | -- | --
0.002 | 0.00363872 | 0.00163872
0.002 | 0.00363872 | 0.00163872
0.0023 | 0.00413805 | 0.00183805
-0.0056 | -0.0067684 | -0.0011684
0.0099 | -0.0164253 | -0.0263253
0.0076 | 0.00757941 | -2.059E-05
-0.0056 | -0.0067684 | -0.0011684
-0.0016 | -0.0015623 | 3.7737E-05
0.0076 | 0.00757941 | -2.059E-05
  |   |  
0.0186 | -0.0049501 |  

The function I used (sorry its not real pretty) is:

```
def strbndTerm(inputString):
    radian = 57.2957795130

    DATA = inputString.split()

    cls1 = DATA[1]
    cls2 = DATA[2]
    cls3 = DATA[3]

    forceTinker_1 = DATA[4]
    k1 = float(forceTinker_1) * ((4.184 / radian) * 10.0)

    forceTinker_2 = DATA[5]
    k2 = float(forceTinker_2) * ((4.184 / radian) * 10.0)
    
    return "<StretchBend class1=\"{}\" class2=\"{}\" class3=\"{}\" k1=\"{:.13f}\" k2=\"{:.13f}\" />"\
        .format(cls1, cls2, cls3, k1, k2)
```



This has worked for the other molecules I've tried so I'm fairly sure it is doing the conversion as intended.

**Comment by peastman** [comment]
Is it possible you're just seeing roundoff error, due to the fact that the structure is very close to the minimum?  Consider the first term listed above, for which Tinker reports 0.002 kJ/mol and OpenMM reports 0.00363872 kJ/mol.  In absolute terms that's a tiny difference.  The relative difference looks large, but only because the energy is tiny.

The first bond length involved in that term has a length of 0.1090 nm, compared to an equilibrium length of 0.1112.  The second bond has a length of 0.15109 compared to an equilibrium length of 0.15247.  And the angle is 1.92737 radians, compared to an equilibrium value of 1.93207.  All of them are very close to equilibrium, leading to a tiny energy, which makes the tiny difference between codes look large.

**Comment by Gareth-elliott** [comment]
Hi @peastman, thanks for your reply :-)

I tried distorting the structure away from the minimum, however the difference persisted.

I kept looking and for some reason if you specify all the options to Tinker's ANALYZE, it only spits out 1 force constant for the stretch-bend terms, however specifying only the P,E,A,D options pulls out both. For most of the molecules (at least in amoeba09.prm) these are equal, although in [#800](https://github.com/openmm/openmm/issues/800)/[#801](https://github.com/openmm/openmm/issues/801) there is discussion/implementation of the general case where they are not equal. It seems as though two of the conversions I did, methanol and ethanol, have unequal k values. 

It seems as though methanol works fine, and the parameters are assigned correctly, however something is going wrong for ethanol:

atom1 | atom2 | atom3 | omm-b1 | omm-b2 | omm-ang | omm-k1 | omm-k2 | omm-E (conv to kcal/mol) | Tinker-b1 | Tinker-b2 | Tinker-ang | Tinker-k1 | Tinker-k2 | Tinker-E | abs(Ediff)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | 5 | 7 | 0.15247 | 0.1112 | 1.90781941 | -3.2861059 | 27.7493388 | -0.0067684 | 1.5247 | 1.112 | 109.31 | 38 | -4.5 | -0.0056 | **0.00116841**
1 | 5 | 8 | 0.15247 | 0.1413 | 1.88321026 | 27.7493388 | 27.7493388 | -0.0015623 | 1.5247 | 1.413 | 107.9 | 38 | 38 | -0.0016 | 3.7737E-05
1 | 5 | 6 | 0.15247 | 0.1112 | 1.90781941 | -3.2861059 | 27.7493388 | -0.0067684 | 1.5247 | 1.112 | 109.31 | 38 | -4.5 | -0.0056 | **0.00116841**
2 | 1 | 5 | 0.1112 | 0.15247 | 1.93207948 | 27.7493388 | -3.2861059 | 0.00363872 | 1.112 | 1.5247 | 110.7 | -4.5 | 38 | 0.002 | **0.00163872**
3 | 1 | 5 | 0.1112 | 0.15247 | 1.93207948 | 27.7493388 | -3.2861059 | 0.00363872 | 1.112 | 1.5247 | 110.7 | -4.5 | 38 | 0.002 | **0.00163872**
4 | 1 | 5 | 0.1112 | 0.15247 | 1.93207948 | 27.7493388 | -3.2861059 | 0.00413805 | 1.112 | 1.5247 | 110.7 | -4.5 | 38 | 0.0023 | **0.00183805**
5 | 8 | 9 | 0.1413 | 0.0967 | 1.86401164 | -3.2861059 | 27.7493388 | -0.0164253 | 1.413 | 0.967 | 106.8 | 38 | -4.5 | 0.0099 | **0.0263253**
6 | 5 | 8 | 0.1112 | 0.1413 | 1.90066356 | -3.2861059 | 27.7493388 | 0.00757941 | 1.112 | 1.413 | 108.9 | -4.5 | 38 | 0.0076 | 2.0594E-05
7 | 5 | 8 | 0.1112 | 0.1413 | 1.90066356 | -3.2861059 | 27.7493388 | 0.00757941 | 1.112 | 1.413 | 108.9 | -4.5 | 38 | 0.0076 | 2.0594E-05

With the pertinent stretch bend terms from amoeba09.prm:

strbnd       37   36   40      -4.50      38.00
strbnd       40   36   40      38.00      38.00
strbnd       36   40   40      38.00      38.00
strbnd       36   40   41      11.50      11.50
strbnd       39   40   36      -4.50      38.00
strbnd       39   40   40      -4.50      38.00
strbnd       40   40   40      18.70      18.70
strbnd       40   40   41      38.00      -4.50

and my converted section in the .xml format:

 <AmoebaStretchBendForce stretchBendUnit="1.0">
  <StretchBend class1="37" class2="36" class3="40" k1="-3.2861059156596" k2="27.7493388433481" />
  <StretchBend class1="40" class2="36" class3="40" k1="27.7493388433481" k2="27.7493388433481" />
  <StretchBend class1="36" class2="40" class3="40" k1="27.7493388433481" k2="27.7493388433481" />
  <StretchBend class1="36" class2="40" class3="41" k1="8.3978262289080" k2="8.3978262289080" />
  <StretchBend class1="39" class2="40" class3="36" k1="-3.2861059156596" k2="27.7493388433481" />
  <StretchBend class1="39" class2="40" class3="40" k1="-3.2861059156596" k2="27.7493388433481" />
  <StretchBend class1="40" class2="40" class3="40" k1="13.6555956939634" k2="13.6555956939634" />
  <StretchBend class1="40" class2="40" class3="41" k1="27.7493388433481" k2="-3.2861059156596" />
 </AmoebaStretchBendForce>

It seems as though the issue is that for some of the terms, the k1 and k2 values are switched. I'm not sure why this is. I thought that it might have been me mis-labelling atoms in my pdb/xml files, but I double checked and I think they are fine, and the other terms in the force field are all equal (stretch-bend is the only thing producing an incorrect energy). The bonds and angles determined by both Tinker and OpenMM match also (see table).

G

**Comment by peastman** [comment]
Here is the code in Tinker that assigns the stretch bend parameters:

https://github.com/TinkerTools/tinker/blob/b829271598e44c0ec8eb78875546d52d19e7ce9e/source/kstrbnd.f#L128-L168

I don't entirely understand what it's doing.  Notice there are a couple of points where it checks `if (ita .le. itc)`.  That is, it assigns the parameters differently depending on whether the first or third atom has a lower atom class.  @jayponder can you shed light on this and what the intended logic is for setting the order of `k1` and `k2`?  Thanks!

**Comment by peastman** [comment]
In processTinkerForceField.py, just before it writes out the `<StretchBend>` tag, try adding these lines:

```python
if int(stretchBend[0]) > int(stretchBend[2]):
   k1, k2 = k2, k1
```

Based on my reading of the Tinker code, I think that should give consistent behavior.

**Comment by jayponder** [comment]
Yes, I think Peter's suggestion above should work. The issue is that the order matters for the stretch-bend parameters. Consider the sample parameter below that is typical of what might be found in a Tinker parameter file:

strbnd          1     2     3     5.0     10.0

This parameter is for a bond angle 1-2-3 formed by atoms with "atom classes" of 1, 2 and 3. The parameter value 5.0 is the strength of the "force constant" for the coupling of the 1-2 bond to the 1-2-3 angle. The 10.0 value is for the coupling of the 2-3 bond to the 1-2-3 angle.

The reason the Tinker code sometimes reverses the force constant values is to handle the case where your molecule is stored as having a 3-2-1 angle, but the parameter file in use has a parameter given for the 1-2-3 angle. In that case the force constants need to be swapped to use that 1-2-3 parameter for the 3-2-1 stored atom order.

**Comment by peastman** [comment]
Thanks!  That makes sense.  In that case, I think the correct change is in ForceField.  As it loops over angles and matches types, it accepts either order:

https://github.com/openmm/openmm/blob/f2cdf70afc3ec4f4f0a4fc3aa5ca4761fc994cf1/wrappers/python/openmm/app/forcefield.py#L4618

Depending on which order passed, we might need to reverse the parameters.  Where it adds the bond

https://github.com/openmm/openmm/blob/f2cdf70afc3ec4f4f0a4fc3aa5ca4761fc994cf1/wrappers/python/openmm/app/forcefield.py#L4656

should be changed to

```python
if type1 in types1 and type3 in types3:
    k1, k2 = self.k1[i], self.k2[i]
else:
    k1, k2 = self.k2[i], self.k1[i]
force.addBond((angle[0], angle[1], angle[2]), (bondAB, bondCB, angleDict['idealAngle']/radian, k1, k2))

**Comment by Gareth-elliott** [comment]
Thanks a lot @peastman and @jayponder :-)

Unfortunately this didn't work:

> In processTinkerForceField.py, just before it writes out the `<StretchBend>` tag, try adding these lines:
> 
> ```python
> if int(stretchBend[0]) > int(stretchBend[2]):
>    k1, k2 = k2, k1
> ```
> 
> Based on my reading of the Tinker code, I think that should give consistent behavior.

However the fix proposed to forcefield.py appears to work for the original ethanol.xml file, and it reproduces the energy predicted by Tinker. I'll open a PR for this (although I'm basically copy and pasting your answer from above).

G

**Comment by peastman** [comment]
Fixed by #3420.
